packages/runtime/container-runtime/src/opLifecycle/README.md

```diff
@@ -66,47 +68,30 @@ As `FlushMode.TurnBased` accumulates ops, it is the most vulnerable to run into
 
-Grouped batching may become problematic for batches which contain reentrant ops. This is the case when changes are made to a DDS inside a DDS 'onChanged' event handler. This means that the reentrant op will have a different reference sequence number than the rest of the ops in the batch, resulting in a different view of the state of the data model.
+The purpose for enabling grouped batching before compression is to eliminate the empty placeholder messages in the chunks. Without grouped batching, a batch with many messages (i.e. more than 4k) will go over the batch size limit just on empty op envelopes alone.
 
-Therefore, when grouped batching is enabled, all batches with reentrant ops are rebased to the current reference sequence number and resubmitted to the data stores so that all ops are in agreement about the state of the data model and ensure eventual consistency.
+Grouped batching is only relevant for `FlushMode.TurnBased`, since `OpGroupingManagerConfig.opCountThreshold` defaults to 2. Grouped batching is opaque to the server and implementations of the Fluid protocol do not need to alter their behavior to support this client feature.
 
-### How to enable
+Grouped Batching can be disabled by setting `IContainerRuntimeOptions.enableGroupedBatching` to `false`.
 
-**This feature is disabled by default**
+See [below](#how-grouped-batching-works) for an example.
 
-If all prerequisites in the previous section are met, enabling the feature can be done via the `IContainerRuntimeOptions` as following:
+### Changes in op semantics
 
-```
-    const runtimeOptions: IContainerRuntimeOptions = {
-        (...)
-        enableGroupedBatching: true,
-        (...)
-    }
-```
+Grouped Batching changed a couple of expectations around message structure and runtime layer expectations. Specifically:
 
-In case of emergency grouped batching can be disabled at runtime, using feature gates. If `"Fluid.ContainerRuntime.DisableGroupedBatching"` is set to `true`, it will disable grouped batching if enabled from `IContainerRuntimeOptions` in the code.
-
-Grouped batching is only relevant for `FlushMode.TurnBased` as it only targets the number of ops in a batch. Grouped batching is opaque to the server and implementations of the Fluid protocol do not need to alter their behavior to support this client feature.
+-   Batch messages observed at the runtime layer no longer match messages seen at the loader layer (i.e. grouped form at loader layer, ungrouped form at runtime layer)
+-   Messages within the same batch now have the same sequence number
+-   Once the ContainerRuntime ungroups the batch, the client sequence numbers on the resulting messages can only be used to order messages with that batch (having the same sequenceNumber)
+-   When grouped batching is enabled, a batch with reentrant ops must rebase its messages to the current reference sequence number and resubmit them to the data stores before flushing.
```

Comment:
I'd probably phrase this section differently. I'd split it into two parts:
- op reentrancy can cause ops in a single JS turn (and thus in a batch) to have different ref seq #.
- The internal requirement - all ops in a batch must have the same ref sequence #, so runtime will rebase ops in a batch with different ref sequence number to satisfy that requirement.

That said, the one part I do not remember (and do not have an answer) - why we are doing rebasing as opposed to closing previous batch (with consistent ref seq #) and start a new batch, when we observe a new op with a different ref seq #. Given that reentrancy is sort of best-effort support case, I do not think we are obligated to ensure all such ops would arrive in a single batch. It's probably too late to change it, and not breaking batch is better, but the cost of getting there was much higher than simple solution, so there must be some explanation for why we chose this path...

Reply:
Good question.  Not sure who would have the answer besides Andrei... even though I could tag him here I won't 😇  @kian-thompson do you know the history of rebasing to ensure consistent refSeq in a batch?